sender/nats.go

```diff
@@ -0,0 +1,181 @@
+	}
+	if n.Servers == "" {
+		return skogul.MissingArgument("Servers")
+	}
+
+	//User Credentials, use either.
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		return fmt.Errorf("Please configure usercreds or nkeyfile.")
+	}
+
+	return nil
+}
+
+func (n *Nats) init() error {
+
+	if n.Encoder.Name == "" {
+		n.Encoder.E = encoder.JSON{}
+	}
+
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.conOpts = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.UserCreds != "" {
+		*n.conOpts = append(*n.conOpts, nats.UserCredentials(n.UserCreds))
+	}
+
+	//Plain text passwords
+	if n.Username != "" && n.Password != "" {
+		if n.TLSClientKey != "" {
+			natsLog.Warnf("Using plain text password over a non encrypted transport!")
+		}
+		*n.conOpts = append(*n.conOpts, nats.UserInfo(n.Username, n.Password))
+	}
+
+	//TLS authentication, Note: Fix selfsigned certificates.
+	if n.TLSClientKey != "" && n.TLSClientCert != "" {
+		cert, err := tls.LoadX509KeyPair(n.TLSClientCert, n.TLSClientKey)
+		if err != nil {
+			n.init_error = fmt.Errorf("error parsing X509 certificate/key pair: %v", err)
+		}
+
+		cp, err := skogul.GetCertPool(n.TLSCACert)
+		if err != nil {
+			n.init_error = fmt.Errorf("Failed to initialize root CA pool")
+		}
+
+		if n.init_error == nil {
+			config := &tls.Config{
+				InsecureSkipVerify: n.Insecure,
+				Certificates:       []tls.Certificate{cert},
+				RootCAs:            cp,
+			}
+			*n.conOpts = append(*n.conOpts, nats.Secure(config))
+		}
+	}
+
+	//NKey auth
+	if n.NKeyFile != "" {
+		opt, err := nats.NkeyOptionFromSeed(n.NKeyFile)
+		if err != nil {
+			natsLog.Fatal(err)
+		}
+		*n.conOpts = append(*n.conOpts, opt)
+	}
+
+	//Log disconnects
+	*n.conOpts = append(*n.conOpts, nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
+		natsLog.WithError(err).Error("Got disconnected!")
+	}))
+	//Log reconnects
+	*n.conOpts = append(*n.conOpts, nats.ReconnectHandler(func(nc *nats.Conn) {
+		natsLog.Info("Reconnected")
+	}))
+	//Always try reconnecting
+	*n.conOpts = append(*n.conOpts, nats.RetryOnFailedConnect(true))
+	//Keep doing reconnects
+	*n.conOpts = append(*n.conOpts, nats.MaxReconnects(-1))
+
+	var err error
+	n.natsCon, err = nats.Connect(n.Servers, *n.conOpts...)
+	if err != nil {
+		n.init_error = fmt.Errorf("Encountered an error while connecting to Nats: %v", err)
+	}
+	return err
+}
+
+func (n *Nats) Send(c *skogul.Container) error {
+	n.once.Do(func() {
+		n.init()
+	})
+	if n.init_error != nil {
+		return n.init_error
+	}
+	for _, m := range c.Metrics {
+		subject := n.Subject
+		a_subject := m.Metadata[n.SubjectAppend]
+		//Append metadata field to subject.
+		if n.SubjectAppend != "" && a_subject.(string) != "" {
+			if !strings.HasSuffix(m.Metadata[n.SubjectAppend].(string), ".") {
+				subject = subject + "."
+			}
+			subject = subject + m.Metadata[n.SubjectAppend].(string)
+		}
+
+		b, err := n.Encoder.E.EncodeMetric(m)
+		if err != nil {
+			natsLog.Warnf("couldn't encode metric: %w", err)
```

Comment:
Please use `.WithErr(err).Warn(...)` here instead. And consider if more detail can be added, such as some metadata, the ID in the list of Metrics, or similar, to make it easier to debug if this happens (this can be logged on another log-level so that it is possible to enable this conditionally while debugging).

Reply:
Good point!